src/utils/graceful_shutdown.rs

```diff
@@ -0,0 +1,208 @@
+//! A class to handle shutdown signals and to shut down system
+//! Actix system has to be running for it to work.
+
+use std::{
+    collections::{BTreeMap, HashSet},
+    mem,
+    time::Duration,
+};
+
+use actix::{
+    self,
+    prelude::{Actor, Context},
+    Addr, AsyncContext, Handler, Message, Recipient, ResponseActFuture, System,
+    WrapFuture,
+};
+
+use tokio::prelude::{
+    future::{self, join_all, Future},
+    stream::*,
+    FutureExt,
+};
+
+use crate::log::prelude::*;
+
+pub type ShutdownMessageResult =
+    Result<Box<(dyn Future<Item = (), Error = ()> + std::marker::Send)>, ()>;
+
+type ShutdownFutureType = dyn Future<Item = Vec<()>, Error = ()>;
+
+#[derive(Debug, Message)]
+#[rtype(result = "ShutdownMessageResult")]
+pub struct ShutdownMessage;
+
+/// Subscribe to exit events, with priority
+#[derive(Message)]
+#[rtype(result = "()")]
+pub struct ShutdownSubscribe {
+    pub priority: u8,
+    pub who: Recipient<ShutdownMessage>,
+}
+
+/// Subscribe to exit events, with priority
+#[derive(Message)]
+#[rtype(result = "()")]
+pub struct ShutdownUnsubscribe {
+    pub priority: u8,
+    pub who: Recipient<ShutdownMessage>,
+}
+
+/// Send this when a signal is detected
+#[cfg(unix)]
+#[derive(Message)]
+#[rtype(result = "Result<(),()>")]
+struct ShutdownSignalDetected(i32);
+
+pub struct GracefulShutdown {
+    /// [`Actor`]s to send message when graceful shutdown
+    recipients: BTreeMap<u8, HashSet<Recipient<ShutdownMessage>>>,
+
+    /// Timeout after which all [`Actors`] will be forced shutdown
+    shutdown_timeout: u64,
+}
+
+impl GracefulShutdown {
+    fn new(shutdown_timeout: u64) -> Self {
+        Self {
+            recipients: BTreeMap::new(),
+            shutdown_timeout,
+        }
+    }
+}
+
+impl Actor for GracefulShutdown {
+    type Context = Context<Self>;
+
+    fn started(&mut self, ctx: &mut Self::Context) {
+        #[cfg(not(unix))]
+        {
+            error!(
+                "Unable to use graceful_shutdown: only UNIX signals are \
+                 supported"
+            );
+            return;
+        }
+        #[cfg(unix)]
+        {
+            use tokio_signal::unix::{
+                Signal, SIGHUP, SIGINT, SIGQUIT, SIGTERM,
+            };
+
+            let sigint_stream = Signal::new(SIGINT).flatten_stream();
+            let sigterm_stream = Signal::new(SIGTERM).flatten_stream();
+            let sigquit_stream = Signal::new(SIGQUIT).flatten_stream();
+            let sighup_stream = Signal::new(SIGHUP).flatten_stream();
+            let signals_stream = sigint_stream
+                .select(sigterm_stream)
+                .select(sigquit_stream)
+                .select(sighup_stream);
```

Comment:
@tyranron,

I cannot find a way to do that.
`SIGINT`, `SIGTERM`, `SIGQUIT` and `SIGHUP` are [i32 constants](https://docs.rs/tokio-signal/0.2.8/tokio_signal/unix/index.html#constants).
When calling
```
let sigint_stream = Signal::new(SIGINT).flatten_stream();
```
we create a future which will resolve after the signal is received. `new` accepts c_int (i32 in libc) and  returns a stream => I don't think there is a way to put a bunch of signals into one `new`.

Then there are a bunch of streams. I can use [select_all](https://docs.rs/futures/0.1.28/futures/future/fn.select_all.html) or a bunch of [select](https://docs.rs/futures/0.1.28/futures/future/trait.Future.html#method.select)s. I think a bunch of selects is fine here, since `select_all` takes an iterator as an argument.

Or I can create a vector of signals to catch, then another one to store signals futures in, then loop over signals vector and push all futures into futures vector, and then pass this future to select_all

Reply:
@nWacky 

> I don't think there is a way to put a bunch of signals into one `new`.

I didn't say that. There is a `.map()` in my pseudo-code, which is responsible to map array of `c_int` into array of futures (or streams).

> Or I can create a vector of signals to catch, then another one to store signals futures in, then loop over signals vector and push all futures into futures vector, and then pass this future to select_all

This is nearly the same what pseudo-code above does. Just instead of using `Vec` (which is heap-allocated), you can use a regular array here. After you have an array of `Stream`s, it's up to you how you will `select` them (and `select_all` is a good choice here, as for me).